src/encoded/types/file.py

```diff
@@ -919,8 +922,32 @@ def build_key(cls, registry, uuid, properties):
 
     @classmethod
     def build_external_creds(cls, registry, uuid, properties):
-        bucket = cls.get_bucket(registry)
+        """ This function is very important in that it determines both the upload
+            and download location of files - so we have two distinct cases to handle.
+
+            It is sometimes the case that we build_external_creds for extra files that
+            may not be in the same bucket as the source file. So we first assume the
+            common (download) case and check for file presence in both buckets and use
+            the one where the file actually is (prioritizing the wfoutput bucket).
+
+            We could also be doing an upload, in which case we will do two head requests
+            that will give us nothing and we should fallback to the bucket associated
+            with the type.
+        """
+        conn = boto3.client('s3')
+        bucket = None
         key = cls.build_key(registry, uuid, properties)
+        # _head_s3 for both files and wfoutput buckets if we are doing a download
+        for b in [registry.settings['file_wfout_bucket'], registry.settings['file_upload_bucket']]:
```

Comment:
If you make the above-proposed change to add an extra parameter, this (untested) code might be better:
```suggestion
        for b in try_buckets:
```

Reply:
I wouldn't mind seeing `b` renamed to `try_bucket` for clarity.